alg/teca_lapse_rate.py

```diff
@@ -0,0 +1,201 @@
+    zmax = 9000
+    z_is_geopotential = True
+    fill_value = 1e20
+    gravity = 9.806
+
+    def set_t_var(self, t_var):
+        """ Sets the temperature variable name. """
+        self.t_var = t_var
+
+    def set_z_var(self, z_var):
+        """ Sets the geopotential height variable name. """
+        self.z_var = z_var
+
+    def set_zs_var(self, zs_var):
+        """ Sets the surface geopotential height variable name. """
+        self.zs_var = zs_var
+
+    def set_zmax(self, zmax):
+        """ Sets the maximum height to use in the lapse rate calculation. """
+        self.zmax = zmax
+
+    def set_geopotential_flag(self, geopotential_flag):
+        """ Flags whether z is geopotential (units m^2/s^2) or height (m) """
+        self.z_is_geopotential = geopotential_flag
+
+    def set_fill_value(self, fill_value):
+        self.fill_value = fill_value
+
+    def get_fill_value(self, fill_value):
+        return self.fill_value
+
+    def get_point_array_names(self):
+        """ Returns the names of the output arrays """
+        return ["lapse_rate"]
+
+    def calculate_lapse_rate(self, t, z, zs, zmax=9000):
+        """ Calculates the mean lapse rate.
+
+            input:
+            ------
+
+                t  : atmospheric temperature [K]
+
+                z  : geopotential height [m]
+
+                zs : surface geopotential height [m]
+
+                zmax : the maximum height to consider in the lapse rate
+                       calculation [m]
+
+            output:
+            -------
+
+                dtdz : the mean lapse rate [K/m]
+
+
+        Uses 1st order finite difference to estimate the lapse rate at all
+        available levels and then averages. Levels below the surface are
+        masked, and levels above zmax are masked.
+
+        It is assumed that the vertical dimension is the first dimension
+        e.g., [level, lat, lon]) for both t and z.
+
+        """
+
+        # check shapes
+        try:
+            assert t.shape == z.shape, "t and z must have the same shape; \
+                   shape(t) = "+str(t.shape)+" but shape(z) = "+str(z.shape)
+        except Exception as e:
+            print(e)
+        try:
+            assert z[0, ...].shape == zs.shape, "All but the first dimension \
+                   of z must match the dimensions of zs; \
+                   shape(z) = "+str(z.shape)+" but shape(zs) = "+str(zs.shape)
+        except Exception as e:
+            print(e)
+
+        # mask temperature and height
+        # for under-topo values and values that are above the max height
+        t_mask = np.where((z >= zs[np.newaxis, ...]) & (z < zmax), t, np.nan)
+        z_mask = np.where((z >= zs[np.newaxis, ...]) & (z < zmax), z, np.nan)
+
+        # calculate the lapse rate between each level
+        dt = np.diff(t_mask, axis=0)
+        dz = np.diff(z_mask, axis=0)
+
+        # calculate the average lapse rate
+        dtdz = np.nanmean((dt/dz), axis=0)
+
+        return dtdz
```

Comment:
In this and some of the other functions you're using numpy all the time (via a global variable `np`). You'll need to select either numpy or cupy in each function and assign to a local variable. Note, what you've done in `execute` is right, you'll need to do this in the other functions as well.